File: 0070-climbing-stairs/0070-climbing-stairs.js
/**
 * @param {number} n
 * @return {number}
 */
 /**
    My Solution 1: DP
  */
var climbStairs = function(n, mem={}) {
    if(n==2) return 2
    if(n==1) return 1
    if(n<=0) return 0
    
    if(!(n in mem)) {
        mem[n] = climbStairs(n-1, mem) + climbStairs(n-2, mem)
    }

    return mem[n]
};

File: 0070-climbing-stairs/0070-climbing-stairs.py
class Solution:
    def climbStairs(self, n: int, mem={}) -> int:
        # if n is 0, then 0 ways
        if n<=0: return 0
        # if n is 1, then only 1 way
        if n==1: return 1
        # if n is 2, then 2 ways, 2 and 1+1
        if n==2: return 2

        # at each step, we can take either 1 or 2 steps
        if n not in mem:
            mem[n] = self.climbStairs(n-1, mem) + self.climbStairs(n-2, mem)

        return mem[n]